graph-builder/src/registry.rs

```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
+) -> Result<Vec<Release>, Error> {
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_path))
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let authenticated_client = tcore
+        .run(authenticate_client(
+            &mut client,
+            format!("repository:{}:pull", &repo),
+        ))
+        .map_err(|e| format_err!("{}", e))?;
+
+    let releases = authenticated_client
+        .get_tags(&repo, None)
+        .and_then(|tag| {
+            trace!("processing: {}:{}", &repo, &tag);
+
+            let tag_clone0 = tag.clone(); // TODO(steveeJ): is there a way to avoid this?
+            let manifest_kind_future = authenticated_client
+                .has_manifest(&repo, &tag, None)
+                .and_then(move |manifest_kind| match manifest_kind {
+                    Some(manifest_kind) => Ok(manifest_kind),
+                    None => {
+                        Err(format!("{}:{} doesn't have a manifest", &repo, &tag_clone0).into())
+                    }
+                })
+                .inspect(|manifest_kind| {
+                    trace!("manifest_kind: {:?}", manifest_kind);
+                });
+
+            let manifest_future = authenticated_client.get_manifest(&repo, &tag);
+
+            let layer_digests_future =
+                manifest_kind_future
+                    .join(manifest_future)
+                    .map(|(manifest_kind, manifest)| match manifest_kind {
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S1Signed => {
+                            let m: dkregistry::v2::manifest::ManifestSchema1Signed =
+                                serde_json::from_slice(manifest.as_slice())?;
+                            Ok((m.get_layers(), tag))
+                        }
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S2 => {
+                            let m: dkregistry::v2::manifest::ManifestSchema2 =
+                            serde_json::from_slice(manifest.as_slice())?;
+                            Ok((m.get_layers(), tag))
+                        }
+                        _ => Err(format_err!("unknown manifest_kind '{:?}'", manifest_kind)),
+                    });
+
+            layer_digests_future
         })
-    }
-    Ok(metadata)
+        .map_err(|e| format_err!("{}", e))
+        .and_then(|layer_digests_tag| {
+            let (layer_digests, tag) = layer_digests_tag?;
+            trace!("tag: {:?} layer_digests: {:?}", &tag, &layer_digests);
+
+            let layer_digests_mapped_to_releases = layer_digests.iter().map(|layer_digest| {
+                let (registry_host, repo, tag) = (registry_host.clone(), repo.clone(), tag.clone());
```

Comment:
Are these clones actually necessary?

Reply:
Yes, because they are consumed by the closures

Reply:
I was able to remove these in the split version of the future chain too.